app/adapters/abstract_adapter.rb

```diff
@@ -0,0 +1,294 @@
+# frozen_string_literal: true
+
+require 'uri'
+require 'httpclient'
+require 'mutex_m'
+
+# KeycloakAdapter adapter to create/update/delete Clients on using the KeycloakAdapter Client Registration API.
+class AbstractAdapter
+  def self.build_client(*)
+    raise NotImplementedError, __method__
+  end
+
+  attr_reader :endpoint
+
+  def initialize(endpoint, authentication: nil)
+    endpoint = EndpointConfiguration.new(endpoint)
+    @http_client = build_http_client(endpoint)
+    @oidc = OIDC.new(endpoint, http_client)
+    @oidc.access_token = authentication if authentication
+    @endpoint = endpoint.issuer
+  end
+
+  def authentication=(value)
+    oidc.access_token = value
+  end
+
+  def authentication
+    oidc.access_token.token
+  end
+
+  def create_client(_)
+    raise NotImplementedError, __method__
+  end
+
+  def read_client(_)
+    raise NotImplementedError, __method__
```

Comment:
Not fan of `NotImplementedError` because it is not really an OS implementation. Ancestor is `ScriptError` and not `StandardError` also.

Preferably `NoMethodError`

```ruby
2.3.8 (main):0 > NotImplementedError.ancestors
=> [NotImplementedError,
 ScriptError,
 Exception,
 ActiveSupport::Dependencies::Blamable,
 Object,
 ERB::Util,
 PP::ObjectMixin,
 JSON::Ext::Generator::GeneratorMethods::Object,
 ActiveSupport::Dependencies::Loadable,
 Kernel,
 BasicObject]
```

Reply:
Good point!